Spigot-Server-Patches/0139-Implement-Profile-api.patch

```diff
@@ -0,0 +1,674 @@
++                profile = getProfileLookup().lookup(name);
++                // Paper end
+             }
+             // Spigot end
+             if (profile == null) {
+@@ -1356,7 +1363,10 @@ public final class CraftServer implements Server {
+                 result = getOfflinePlayer(new GameProfile(UUID.nameUUIDFromBytes(("OfflinePlayer:" + name).getBytes(Charsets.UTF_8)), name));
+             } else {
+                 // Use the GameProfile even when we get a UUID so we ensure we still have a name
+-                result = getOfflinePlayer(profile);
++                // Paper start
++                result = new CraftOfflinePlayer(this, profile);
++                offlinePlayers.put(result.getUniqueId(), result);
++                // Paper end
+             }
+         } else {
+             offlinePlayers.remove(result.getUniqueId());
+@@ -1373,7 +1383,7 @@ public final class CraftServer implements Server {
+         if (result == null) {
+             result = offlinePlayers.get(id);
+             if (result == null) {
+-                result = new CraftOfflinePlayer(this, new GameProfile(id, null));
++                result = new CraftOfflinePlayer(this, getProfileLookup().lookup(id)); // Paper - do lookup
+                 offlinePlayers.put(id, result);
+             }
+         } else {
+@@ -1415,7 +1425,7 @@ public final class CraftServer implements Server {
+ 
+         for (JsonListEntry entry : playerList.getProfileBans().getValues()) {
+             result.add(getOfflinePlayer((GameProfile) entry.getKey()));
+-        }        
++        }
+ 
+         return result;
+     }
+@@ -1844,6 +1854,14 @@ public final class CraftServer implements Server {
+         return spigot;
+     }
+ 
++    // Paper start - uuid api
++    private final ProfileLookup lookup = new MojangLookup();
++
++    @Override
++    public ProfileLookup getProfileLookup() {
++        return lookup;
++    }
++    // Paper end
+     // Paper start
+     @Override
+     public void reloadPermissions() {
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+index 4378795..901b0db 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+@@ -59,6 +59,9 @@ import org.bukkit.metadata.MetadataValue;
+ import org.bukkit.plugin.Plugin;
+ import org.bukkit.plugin.messaging.StandardMessenger;
+ import org.bukkit.scoreboard.Scoreboard;
++// PaperSpigot start
++import com.destroystokyo.paper.profile.PlayerProfile;
++// PaperSpigot end
+ 
+ @DelegateDeserialization(CraftOfflinePlayer.class)
+ public class CraftPlayer extends CraftHumanEntity implements Player {
+@@ -83,13 +86,20 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+         firstPlayed = System.currentTimeMillis();
+     }
+ 
+-    public GameProfile getProfile() {
```

Comment:
Bad idea renaming this - change the API method to `getPlayerProfile` or something.